scripts/generateTranslations.ts

```diff
@@ -589,55 +571,333 @@ class TranslationGenerator {
+                return nodeToUse;
+            }
+
+            // Extract complex expressions from the transformed node (which already has translations applied)
+            const translatedComplexExpressions = new Map<number, ts.Expression>();
+
+            // Use the transformed expressions - they'll have nested translations applied for complex expressions
+            // and be identical to originals for simple expressions
+            const transformedTemplateNode = transformedNode && ts.isTemplateExpression(transformedNode) ? transformedNode : node;
+            for (let i = 0; i < node.templateSpans.length; i++) {
+                const originalExpression = node.templateSpans[i].expression;
+                const transformedExpression = transformedTemplateNode.templateSpans[i].expression;
+
+                if (!this.isSimpleExpression(originalExpression)) {
+                    const hash = hashStr(originalExpression.getText());
+                    translatedComplexExpressions.set(hash, transformedExpression);
+                }
+            }
+
+            // Build the translated template expression, referencing the translated template spans as necessary
+            return this.stringToTemplateExpression(translatedTemplate, translatedComplexExpressions);
+        }
+
+        return nodeToUse;
+    }
+
+    /**
+     * Extract translated code strings from a transformed AST for the specified paths.
+     */
+    private extractTranslatedNodes(sourceFile: ts.SourceFile, translatedCodeMap: Map<string, string>): void {
+        const visitWithPath = (node: ts.Node, currentPath = '') => {
+            // Only extract code strings for exact paths in our sets (not hierarchical matches)
+            const isAddedPath = this.pathsToAdd.has(currentPath as TranslationPaths);
+            const isModifiedPath = this.pathsToModify.has(currentPath as TranslationPaths);
+
+            if ((isAddedPath || isModifiedPath) && ts.isPropertyAssignment(node)) {
+                if (!node.initializer) {
+                    throw new Error('Found a dangling property without an initializer in a translation object. This should never happen.');
+                }
+
+                // extract the value (property initializer) as code string
+                const codeString = tsPrinter.printNode(ts.EmitHint.Expression, node.initializer, sourceFile);
+                translatedCodeMap.set(currentPath, codeString);
+
+                return; // Stop recursing into children
+            }
+
+            // Continue traversing children, updating path for property assignments
+            node.forEachChild(TSCompilerUtils.createPathAwareVisitor(visitWithPath, currentPath));
         };
+
+        visitWithPath(sourceFile);
+    }
+
+    /**
+     * Extract removed paths by traversing the old version of en.ts.
+     */
+    private extractRemovedPaths(removedLines: Set<number>): void {
```

Comment:
I'm not sure that I understand why we need to traverse the old version of en.ts to get the removed paths. Could you explain that please?

Reply:
Discussed this 1:1 with Rory, I get why we do this now.